### PR TITLE
refactor(shorebird_cli): introduce ArtifactManager as home for common artifact operations

### PR DESCRIPTION
**Pull request by bryanoltman**
## Description

Introduces the `ArtifactManager` to handle common artifact operations, such a downloading, diffing, and unzipping.

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 1]
```diff
@@ -22,7 +23,10 @@ import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
 /// {@endtemplate}
 class PreviewCommand extends ShorebirdCommand {
   /// {@macro preview_command}
-  PreviewCommand() {
+  PreviewCommand({
+    http.Client? httpClient,
+  }) : _httpClient = httpClient ??
+            retryingHttpClient(LoggingClient(httpClient: http.Client())) {
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart:29]: might be nice to have a scoped httpClient for this type of thing

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 2]
```diff
@@ -148,14 +154,17 @@ class PreviewCommand extends ShorebirdCommand {
...
-      appId: appId,
-      release: release,
-      platform: platform,
-      extension: 'aab',
+    final aabFile = File(
+      getArtifactPath(
+        appId: appId,
+        release: release,
+        platform: platform,
+        extension: 'aab',
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart:162]: isn't `extension` a keyword? We might want to rename that property.

File: packages/shorebird_cli/lib/src/commands/preview_command.dart [hunk 3]
```diff
@@ -148,14 +154,17 @@ class PreviewCommand extends ShorebirdCommand {
...
+      ),
     );
 
-    if (!File(aabPath).existsSync()) {
+    if (!aabFile.existsSync()) {
+      aabFile.createSync(recursive: true);
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/commands/preview_command.dart:167]: why is this needed? Shouldn't this be handled by artifact manager's downloadFile?

File: packages/shorebird_cli/lib/src/artifact_manager.dart [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  /// Extracts the [zipFile] to the [outputPath] directory.
+  Future<void> extractZip({
+    required File zipFile,
+    required Directory outputPath,
+  }) async {
+    await Isolate.run(() async {
+      final inputStream = InputFileStream(zipFile.path);
+      final archive = ZipDecoder().decodeBuffer(inputStream);
```
Comment (felangel) [inline packages/shorebird_cli/lib/src/artifact_manager.dart:85]: probably also want to be able to mock the zip decoder to speed up test runs but doesn't have to be done in this PR

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 3 above)
Comment (bryanoltman) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart:167]: Good call, fixed.

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 2 above)
Comment (bryanoltman) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart:162]: This wasn't introduced in this PR, but seems fine aside from the syntax highlighting?

File: packages/shorebird_cli/lib/src/commands/preview_command.dart (hunk 1 above)
Comment (bryanoltman) [reply packages/shorebird_cli/lib/src/commands/preview_command.dart:29]: Agreed, filed https://github.com/shorebirdtech/shorebird/issues/1310

**Review by felangel — APPROVED** [review]
